### PR TITLE
Retry loading of mix file

### PR DESCRIPTION
**Pull request by RoosterDragon**
Binaries built by mono and running on Windows exhibit a bug where the parsing the the mix file will occasionally produce garbage results. Retrying the operation is sufficient to mitigate the issue.

---

This should prevent any more crashes due to #2441 for now, until there is time for somebody to investigate the root cause of the problem and have a proper fix implemented.

(Many thanks to @pi99y for their tireless efforts in helping to investigate the issue)

### CODE REVIEW

File: OpenRA.Game/FileSystem/MixFile.cs [hunk 1]
```diff
@@ -56,7 +56,27 @@ public MixFile(string filename, PackageHashType type, int priority)
 			this.priority = priority;
 			this.type = type;
 
-			s = GlobalFileSystem.Open(filename);
+			// HACK: Binaries built by mono that run on Windows exhibit an issue where parsing this file sometimes
+			// returns garbage results. This usually manifests in an exception trying to build the index as there will
+			// be many duplicate entries. Retrying is generally sufficient for a successful parse.
+			const int Attempts = 3;
+			for (var i = 1; i <= Attempts; i++)
```
Comment (pchote) [inline OpenRA.Game/FileSystem/MixFile.cs]: Minor style nit: `i = 0; i < Attempts;` please.

File: OpenRA.Game/FileSystem/MixFile.cs [hunk 2]
```diff
@@ -56,7 +56,27 @@ public MixFile(string filename, PackageHashType type, int priority)
...
+			{
+				try
+				{
+					LoadMixFile(filename, out s, out index, out dataStart);
```
Comment (pchote) [inline OpenRA.Game/FileSystem/MixFile.cs]: You'll need to break on success.

Comment (RoosterDragon) [inline OpenRA.Game/FileSystem/MixFile.cs]: Derp - done.

File: OpenRA.Game/FileSystem/MixFile.cs (hunk 1 above)
Comment (RoosterDragon) [inline OpenRA.Game/FileSystem/MixFile.cs]: This would require ```i == Attempts - 1``` below. Are you sure?

Comment (pchote) [inline OpenRA.Game/FileSystem/MixFile.cs]: I guess this is fine, then.  Both options are about equally style-nitty.

Comment (RoosterDragon) [inline OpenRA.Game/FileSystem/MixFile.cs]: I'll leave it as-is for now.

**Comment by Mailaender** [comment]
:+1: Still works fine on Linux. I am actually fine with "Closes https://github.com/OpenRA/OpenRA/issues/2441" as the problem may be solved in practice although this clearly is the brute force method.

**Comment by RoosterDragon** [comment]
Probably best to leave it open for now - this isn't a 100% guaranteed fix.

**Comment by pchote** [comment]
I agree - lets keep this on hold for now, and we can always apply it later in the release cycle if we decide we don't want to debug this further.

**Comment by RoosterDragon** [comment]
To clarify my comment: I still think this is worth merging as it is likely (although not guaranteed) to prevent the crashes - but we should keep the issue open until:
- We're happy this is actually helping.
- We either fix this properly, or decide it isn't worth the effort to debug and fix properly.

**Comment by penev92** [comment]
I'm going to say it's probably not worth the time and effort to look for a proper fix.

**Comment by reaperrr** [comment]
I think unless someone manages to fix the bug properly before that, we should definitely merge this before the next release, a work-around is still better than no fix at all.
At some point we should rather ditch using .mix files directly and instead extract and repack files into .zips, until then this should be good enough.

**Comment by pi99y** [comment]
I agree with @reaperrr, please include it in the next playtest/release. Thanks.

**Comment by pchote** [comment]
I have tracked down the actual problem (a miscompile related to the `fixed` statement), which confirms that this does not fix the problem.  I will file a PR that fixes the underlying problem.

**Comment by RoosterDragon** [comment]
Sweet.

**Comment by GraionDilach** [comment]
This should be removed from the Future milestone as well?

**Pull request closed without merging**